### ISSUE TITLE
build problem with libtool

### ISSUE THREAD

**Issue by kmuehlbauer** [conda-forge/gdal-feedstock#81]
As experienced in #79, there is currently a problem building gdal python bindings using libtool.

I contacted the gdal-dev list for help in this matter: https://lists.osgeo.org/pipermail/gdal-dev/2016-June/044658.html.

**Comment by gillins** [comment]
I'm now getting a failure on OSX when trying to import `gdal`. Seems the patch to `libgdal.dylib` has been hardcoded:
```
otool -L /Users/sam/miniconda3/envs/myenv/lib/python3.5/site-packages/osgeo/_gdal.cpython-35m-darwin.so
/Users/sam/miniconda3/envs/myenv/lib/python3.5/site-packages/osgeo/_gdal.cpython-35m-darwin.so:
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.0.0)
	/Users/travis/miniconda3/conda-bld/work/gdal-2.1.0/libgdal.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/./libstdc++.6.dylib (compatibility version 7.0.0, current version 7.19.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1197.1.1)
```

**Comment by jakirkham** [comment]
...and it seems to still be linked to `libstdc++`. 😒

**Comment by gillins** [comment]
Yep sure is... Not sure if that is related to `libtool`. 

I have been thinking we need a better test for library linking. Would it be possible to add tests for `libstdc++` and libraries under `/Users` etc in an automated way to the build system?

**Comment by jakirkham** [comment]
Certainly worth thinking about.

FWIW, PR ( https://github.com/conda-forge/toolchain-feedstock/pull/9 ) got accepted so the `toolchain` now used 10.9. Maybe this problem already goes away because of that. 🍀

**Comment by gillins** [comment]
That would be good if it does. I'm just testing on my Mac and will report back...

**Comment by gillins** [comment]
With the latest `toolchain`, `libstdc++` goes away, but the absolute path to the build location of `libgdal.dylib` still gets inserted...

**Comment by jakirkham** [comment]
Maybe it is worth trying `libtool` again?

**Comment by jakirkham** [comment]
...or maybe one could try `export LDFLAGS="${LDFLAGS} -Wl,-rpath,$PREFIX/lib"`.

**Comment by jakirkham** [comment]
(after activating the `toolchain` of course)

**Comment by gillins** [comment]
With `libtool` I get the classic error:
```
ibtool: compile: cannot determine name of library object from `build/lib.macosx-10.6-x86_64-3.5/osgeo/_gdal.cpython-35m-darwin.so'
```
I'll try the `rpath` thing next.

**Comment by gillins** [comment]
The `rpath` didn't help...

**Comment by gillins** [comment]
@danclewley How are you guys building `gdal` on OSX these days?

**Comment by kmuehlbauer** [comment]
gdal is built `--without-libtool` at their github mirror: https://github.com/OSGeo/gdal/blob/trunk/.travis.yml

**Comment by gillins** [comment]
Do they have anything for building on OSX?

**Comment by danclewley** [comment]
@gillins the latest `gdal` recipie we have been using is: [here](https://bitbucket.org/danclewley/conda-recipes/src/ed97c0242661b7489972f5f5a2cab6dd39a66714/gdal/build.sh?at=default&fileviewer=file-view-default). We hadn't been specifying an option for libtool and were setting `--disable-rpath` on both Linux and OS X. I'm not sure these are the best practices but were working for us.

**Comment by kmuehlbauer** [comment]
> Do they have anything for building on OSX?

Not to my knowledge :confused:

**Comment by kmuehlbauer** [comment]
@gillins You have been [here](https://trac.osgeo.org/gdal/wiki/BuildingOnMac) before? There are instructions for building gdal on osx. 

I didn't spot anything related to our particular problem, but the `configure` with `--disable-static` and `--with-macosx-framework`. Also they put `CFLAGS`, `CXXFLAGS` and `LDFLAGS` declarations at the end of the `configure`-call. Not sure, if this would change anythings at all.

**Comment by gillins** [comment]
Hadn't seen that - thanks @kmuehlbauer. I tried specifying `--with-macosx-framework` but that didn't help. I don't understand the various `CFLAGS` etc, but specifying them on the command line instead on `export`ing them shouldn't make any difference.
Just about to try setting `--disable-rpath` as suggested by @danclewley

**Comment by gillins** [comment]
That didn't seem to help. Just to recap on OSX: when I use `libtool` I get an error building the Python bindings. Without `libtool` the Python binding (`_gdal.cpython-35m-darwin.so`) ends up with the build location of `libgdal` in the `rpath`.

I've just noticed that the Python binding is called `_gdal.cpython-35m-darwin.so` rather than **.dylib** which can't be good.

I've also tried **not** building the Python bindings via `configure` and doing it separately (cd swig/python etc), but it still ends with with the path hardcoded. Anyone have any other ideas?

**Comment by gillins** [comment]
When I build the Python bindings via Python (`cd swig/python;$PYTHON setup.py install` etc), this command gets executed:
```
clang++ -bundle -undefined dynamic_lookup -L/Users/sam/miniconda3/envs/_build/lib -headerpad_max_install_names -headerpad_max_install_names -mmacosx-version-min=10.9 -lc++ -L/Users/sam/miniconda3/envs/_build/lib -Wno-error=unused-command-line-argument-hard-error-in-future -headerpad_max_install_names -mmacosx-version-min=10.9 -lc++ -L/Users/sam/miniconda3/envs/_build/lib -mmacosx-version-min=10.9 -m64 -I/Users/sam/miniconda3/envs/_build/include -Wno-error=unused-command-line-argument-hard-error-in-future build/temp.macosx-10.6-x86_64-3.5/extensions/gdal_wrap.o -L../../.libs -L../../ -L/Users/sam/miniconda3/envs/_build/lib -L/Users/sam/miniconda3/conda-bld/work/gdal-2.1.0/lib -lgdal -o build/lib.macosx-10.6-x86_64-3.5/osgeo/_gdal.cpython-35m-darwin.so
```
But this is the result:
```
$ otool -L lib/python3.5/site-packages/osgeo/_gdal.cpython-35m-darwin.so 
lib/python3.5/site-packages/osgeo/_gdal.cpython-35m-darwin.so:
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.1.0)
	/Users/sam/miniconda3/conda-bld/work/gdal-2.1.0/libgdal.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
```

**Comment by kmuehlbauer** [comment]
@gillins Seems really problematic things going on and a bunch of red herrings too.

Take a look at this (rather) [old mapnik post](https://github.com/mapnik/mapnik/issues/1683#issuecomment-21804664) regarding `gdal` and `libtool`. In the issue is also some more insights in XCode and `libc++` stuff.

I can't test this locally without a mac. But for the `/bin/bash` swig/python-problem, I discovered that this runs smoothly if you give  `CC` and `CXX` explicitely with the configure call:

```bash
./configure CC=$CC \
            CXX=$CXX \
            --prefix=$PREFIX \
            --with-hdf4=$PREFIX \
            --with-hdf5=$PREFIX \
            --with-xerces=$PREFIX \
            --with-netcdf=$PREFIX \
            --with-geos=$PREFIX/bin/geos-config \
            --with-kea=$PREFIX/bin/kea-config \
            --with-static-proj4=$PREFIX \
            --with-libz=$PREFIX \
            --with-png=$PREFIX \
            --with-jpeg=$PREFIX \
            --with-libjson-c=$PREFIX \
            --with-expat=$PREFIX \
            --with-freexl=$PREFIX \
            --with-libtiff=$PREFIX \
            --with-xml2=$PREFIX \
            --with-openjpeg=$PREFIX \
            --with-spatialite=$PREFIX \
            --with-pg=$PREFIX/bin/pg_config \
            --with-sqlite3=$PREFIX \
            --with-curl \
            --with-python \
            $OPTS
```

This builds fine locally. Checking this in #85.

**Comment by gillins** [comment]
Thanks for the post. Yes we did come across the problem with `libtool` dropping the `-stdlib=libc++` problem, hence why we have:
```
export CXX="${CXX} -stdlib=libc++"
```
For OSX in `build.sh`. I'm trying #85 on my Mac now, will report back there with the results.

**Comment by kmuehlbauer** [comment]
Problem seems to be having CC and CXX exported in the environment and using `configure` and `libtool` building gdal.

Here is another post on the [gdal-dev list](https://lists.osgeo.org/pipermail/gdal-dev/2016-July/044678.html)

**Comment by ocefpaf** [comment]
@kmuehlbauer thanks for digging this up (and sorry for all the trouble you went to find it). I believe we are reaching a solution in #85

**Comment by kmuehlbauer** [comment]
solved in iteration of #85 and #90
